File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        if len(nums) == 3:
            return max(nums[0] + nums[2], nums[1])
        values = [0] * len(nums)
        values[-1] = nums[-1]
        values[-2] = nums[-2]
        values[-3] = nums[-3] + nums[-1]
        
        index = len(nums) - 4
        while index >= 0:
            values[index] = nums[index] + max(values[index + 2], values[index + 3])
            index -= 1
        print(values)

        return max(values[0], values[1])

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        int product = 1;
        boolean hasZero = false;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                if (hasZero) {
                    product = 0;
                }
                hasZero = true;
            }
            else {
                product *= nums[i];
            }
        }

        for (int i = 0; i < nums.length; i++) {
            if (hasZero) {
                if (nums[i] != 0) {
                    res[i] = 0;
                }
                else {
                    res[i] = product;
                }
            }
            else {
                res[i] = product / nums[i];
            }
        }

        return res;
    }
}  
